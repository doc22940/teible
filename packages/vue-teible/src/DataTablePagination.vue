<template>
  <nav :class="theme.datatable__pagination">
    <a
      :disabled="reachedFirst" :class="[theme.datatable__plink, theme.datatable__pprevious]" href="#"
      aria-label="Previous" @click.prevent="load(page-1)"
    >
      <span aria-hidden="true">&laquo;</span>
    </a>
    <a
      :disabled="reachedLast" :class="[theme.datatable__plink, theme.datatable__pnext]" href="#"
      aria-label="Next" @click.prevent="load(page+1)"
    >
      <span aria-hidden="true">&raquo;</span>
    </a>

    <ul :class="theme.datatable__plist">
      <li v-for="(p, index) in pages" :key="index" :class="theme.datatable__pitem">
        <a
          :disabled="p.disabled" :class="{
            [theme.datatable__plink]: true,
            [theme['datatable__plink--active']]: isActive(p),
          }" href="#"
          @click.prevent="load(p.value, p.disabled)"
        >{{ p.value }}</a>
      </li>
    </ul>
  </nav>
</template>
<script>
import { paginate } from 'teible'

export default {
  name: 'DataTablePagination',
  props: {
    total: {
      type: Number,
      required: true
    },
    perPage: {
      type: Number,
      required: true
    },
    page: {
      type: Number,
      required: true
    }
  },
  computed: {
    pages () {
      return paginate(this.page, this.totalPages, 3, 2)
    },
    totalPages () {
      return Math.ceil(this.total / (this.perPage || 1))
    },
    reachedFirst () {
      return this.page === 1
    },
    reachedLast () {
      return this.page >= this.totalPages
    },
    theme () {
      return this.$theme()
    }
  },
  methods: {
    isActive (page) {
      return !page.disabled && this.page === page.value
    },
    load (page, disabled) {
      if (disabled) {
        return
      }

      if (page < 1) {
        return
      }

      if (page > this.totalPages) {
        return
      }

      this.$emit('update:page', page)
    }
  },
  inject: ['$theme']
}
</script>
