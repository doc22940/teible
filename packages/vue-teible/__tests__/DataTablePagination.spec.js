import { shallowMount } from '@vue/test-utils'
import idObj from 'identity-obj-proxy'
import DataTablePagination from '../src/DataTablePagination.vue'

describe('DataTablePagination', () => {
  // wrapper.setProps lines are to mock two-way binding

  it(`renders correctly`, () => {
    let wrapper = shallowMount(DataTablePagination, {
      propsData: {
        total: 10,
        page: 2,
        perPage: 3
      },
      provide: () => ({
        $theme: () => idObj
      })
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it(`renders correctly with invalid perPage`, () => {
    let wrapper = shallowMount(DataTablePagination, {
      propsData: {
        total: 10,
        page: 1,
        perPage: 0
      },
      provide: () => ({
        $theme: () => idObj
      })
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it(`renders correctly when total change`, () => {
    let wrapper = shallowMount(DataTablePagination, {
      propsData: {
        total: 10,
        page: 4,
        perPage: 3
      },
      provide: () => ({
        $theme: () => idObj
      })
    })
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({
      total: 9
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it(`emits update:page events`, () => {
    let wrapper = shallowMount(DataTablePagination, {
      propsData: {
        total: 10,
        perPage: 3,
        page: 2
      },
      provide: () => ({
        $theme: () => idObj
      })
    })

    wrapper.find('.datatable__pprevious').trigger('click')
    expect(wrapper.emitted()).toEqual({ 'update:page': [[1]] })
    wrapper.setProps({
      'page': 1
    })

    // < > [1] 2 3 4

    /*
      Nothing emitted since we reached the first page
    */
    wrapper.find('.datatable__pprevious').trigger('click')
    expect(wrapper.emitted()).toEqual({ 'update:page': [[1]] })

    wrapper.findAll('.datatable__plink').at(5).trigger('click')
    expect(wrapper.emitted()).toEqual({ 'update:page': [[1], [4]] })
    wrapper.setProps({
      'page': 4
    })

    /*
      Nothing emitted since we reached the last page
    */
    wrapper.find('.datatable__pnext').trigger('click')
    expect(wrapper.emitted()).toEqual({ 'update:page': [[1], [4]] })
  })

  it(`does not emit update:page event when clicking on the three-dots button`, () => {
    let wrapper = shallowMount(DataTablePagination, {
      propsData: {
        total: 10,
        page: 1,
        perPage: 1
      },
      provide: () => ({
        $theme: () => idObj
      })
    })
    expect(wrapper.html()).toMatchSnapshot() // < > [1] 2 3 ... 9 10

    wrapper.findAll('.datatable__plink').at(3).trigger('click')
    expect(wrapper.emitted()).toEqual({ 'update:page': [[2]] }) // < > 1 [2] 3 ... 9 10
    wrapper.setProps({
      'page': 2
    })

    // Clicking on the [...]
    wrapper.findAll('.datatable__plink').at(5).trigger('click')
    expect(wrapper.emitted()).toEqual({ 'update:page': [[2]] }) // < > 1 [2] 3 ... 9 10

    wrapper.findAll('.datatable__plink').at(7).trigger('click')
    expect(wrapper.emitted()).toEqual({ 'update:page': [[2], [10]] }) // < > 1 2 3 ... 9 [10]
    wrapper.setProps({
      'page': 10
    })
  })
})
