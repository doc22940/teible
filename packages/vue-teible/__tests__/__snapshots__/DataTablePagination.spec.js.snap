// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTablePagination does not emit update:page event when clicking on the three-dots button 1`] = `
<nav class="datatable__pagination"><a disabled="disabled" href="#" aria-label="Previous" class="datatable__plink datatable__pprevious"><span aria-hidden="true">«</span></a> <a href="#" aria-label="Next" class="datatable__plink datatable__pnext"><span aria-hidden="true">»</span></a>
  <ul class="datatable__plist">
    <li class="datatable__pitem"><a href="#" class="datatable__plink datatable__plink--active">1</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink">2</a></li>
    <li class="datatable__pitem"><a disabled="disabled" href="#" class="datatable__plink">...</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink">9</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink">10</a></li>
  </ul>
</nav>
`;

exports[`DataTablePagination renders correctly 1`] = `
<nav class="datatable__pagination"><a href="#" aria-label="Previous" class="datatable__plink datatable__pprevious"><span aria-hidden="true">«</span></a> <a href="#" aria-label="Next" class="datatable__plink datatable__pnext"><span aria-hidden="true">»</span></a>
  <ul class="datatable__plist">
    <li class="datatable__pitem"><a href="#" class="datatable__plink">1</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink datatable__plink--active">2</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink">3</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink">4</a></li>
  </ul>
</nav>
`;

exports[`DataTablePagination renders correctly when total change 1`] = `
<nav class="datatable__pagination"><a href="#" aria-label="Previous" class="datatable__plink datatable__pprevious"><span aria-hidden="true">«</span></a> <a disabled="disabled" href="#" aria-label="Next" class="datatable__plink datatable__pnext"><span aria-hidden="true">»</span></a>
  <ul class="datatable__plist">
    <li class="datatable__pitem"><a href="#" class="datatable__plink">1</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink">2</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink">3</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink datatable__plink--active">4</a></li>
  </ul>
</nav>
`;

exports[`DataTablePagination renders correctly when total change 2`] = `
<nav class="datatable__pagination"><a href="#" aria-label="Previous" class="datatable__plink datatable__pprevious"><span aria-hidden="true">«</span></a> <a disabled="disabled" href="#" aria-label="Next" class="datatable__plink datatable__pnext"><span aria-hidden="true">»</span></a>
  <ul class="datatable__plist">
    <li class="datatable__pitem"><a href="#" class="datatable__plink">1</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink">2</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink">3</a></li>
  </ul>
</nav>
`;

exports[`DataTablePagination renders correctly with invalid perPage 1`] = `
<nav class="datatable__pagination"><a disabled="disabled" href="#" aria-label="Previous" class="datatable__plink datatable__pprevious"><span aria-hidden="true">«</span></a> <a href="#" aria-label="Next" class="datatable__plink datatable__pnext"><span aria-hidden="true">»</span></a>
  <ul class="datatable__plist">
    <li class="datatable__pitem"><a href="#" class="datatable__plink datatable__plink--active">1</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink">2</a></li>
    <li class="datatable__pitem"><a disabled="disabled" href="#" class="datatable__plink">...</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink">9</a></li>
    <li class="datatable__pitem"><a href="#" class="datatable__plink">10</a></li>
  </ul>
</nav>
`;
